### PR TITLE
My Site Dashboard - Phase 2 - Fix duplicate Error Snackbar

### LINKED ISSUE #15769: My Site Dashboard - Phase 2 - Fix duplicate Error Snackbar
Parent: #15198

### Expected behavior

When a site is switched and there's a dashboard update error, then the `snackbar` should be shown only once.

### Actual behavior

When a site is switched and there's a dashboard update error, then the `snackbar` is shown twice.

### Steps to reproduce the behavior

- Install the app and let at least one dashboard card appear for a site. 
- Inside `FluxC` code, return an error `CardsResult(CardsError(CardsErrorType.GENERIC_ERROR))` from `CardsStore.storeCards`.
- Use `localFluxCPath` inside `WordPress-Android` app.
- Reinstall the app. 
- Notice that the error `snackbar` is shown.
- Switch between sites and return to this same site.
- Notice that the `snackbar` is shown twice. 

##### Tested on [device], Android [version], WPAndroid [version]
Tested on [Pixel 5], Android [12], WPAndroid [18.9]

### PR DESCRIPTION
**Pull request by ashiagr**
Fixes #15769

This PR fixes the duplicate `Error Snackbar` issue on a site switch.

https://user-images.githubusercontent.com/1405144/148212370-56f32bc5-3817-4e78-b51b-835d0bafc857.mp4

The issue occurred because we refreshed the `My Site Sources` inside `onResume` just after building them on a site switch. For the dashboard cards source, since both build and refresh returned a snackbar error, we started seeing a duplicate snackbar. This PR
- Prevents refreshing all sources after build by invoking `MySiteSourceManager.onResume` with a value `true` every time a site is selected (it can be on the initial fragment resume or site switch). - faf9c25
- Displays a progress bar whenever a source build is invoked (or indirectly a site is selected). - fefc5db

To test:

- Install the app and let at least one dashboard card appear for a site. 
- Inside `FluxC` code, return an error `CardsResult(CardsError(CardsErrorType.GENERIC_ERROR))` from `CardsStore.storeCards`.
- Use `localFluxCPath` inside `WordPress-Android` app.
- Reinstall the app. 
- Notice that the error `snackbar` is shown.
- Switch to a different site, then return to the previous site (for other sites error card might be displayed instead of the snackbar with no dashboard cards entries the db).
- Notice that a progress indicator is shown when a site is switched and the error `snackbar` is shown only once for the site.

## Regression Notes
1. Potential unintended areas of impact
The initial resume scenario (see faf9c25) should work as before. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually. Updated existing unit tests.

3. What automated tests I added (or what prevented me from doing so)
Updated existing unit tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartCardSource.kt [hunk 1]
```diff
@@ -21,7 +21,7 @@ class QuickStartCardSource @Inject constructor(
     private val quickStartUtilsWrapper: QuickStartUtilsWrapper,
     private val selectedSiteRepository: SelectedSiteRepository
 ) : MySiteRefreshSource<QuickStartUpdate> {
-    override val refresh = MutableLiveData(true)
+    override val refresh = MutableLiveData(false)
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartCardSource.kt:24]: 1. Question (❓): As we are only calling `refresh()` during `build(...)` if a specific condition is met, will that work every time? 
2. Question (❓): Unrelated to this class, but how about `CurrentAvatarSource` and setting the `refresh` live data to `false` there as well, then using the `refresh()` instead of `result.refreshData()`. Is that not updated as well for a reason?

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/dynamiccards/DynamicCardsSource.kt [hunk 2]
```diff
@@ -17,14 +17,14 @@ class DynamicCardsSource
     private val dynamicCardStore: DynamicCardStore,
     private val selectedSiteRepository: SelectedSiteRepository
 ) : MySiteRefreshSource<DynamicCardsUpdate> {
-    override val refresh = MutableLiveData(true)
+    override val refresh = MutableLiveData(false)
 
     override fun build(coroutineScope: CoroutineScope, siteLocalId: Int): LiveData<DynamicCardsUpdate> {
         val data = MediatorLiveData<DynamicCardsUpdate>()
-        data.refreshData(coroutineScope, siteLocalId)
         data.addSource(refresh) {
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/dynamiccards/DynamicCardsSource.kt]: Minor (🔍): This can become a one-liner as well, just like the rest.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 3]
```diff
@@ -126,6 +126,10 @@ class MySiteViewModel @Inject constructor(
     private val _activeTaskPosition = MutableLiveData<Pair<QuickStartTask, Int>>()
     private val _onShowSwipeRefreshLayout = MutableLiveData<Event<Boolean>>()
 
+    /* Capture and track the site selected event so we can circumvent refreshing sources on resume
+       as they're already built on site select. */
+    private var _isSiteSelected = false
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Minor (🔍): The `_` backing field prefix is not needed here, consider removing it to avoid confusion about it being a live data field.

**Review by ParaskP7 — COMMENTED** [review]
👋 @ashiagr !

I have reviewed and tested this PR as per the instructions, everything works as expected, great job! 🌟 🌟 

I think this is a good approach and I would be personally okay with it if @zwarm also feels the same. Just let me know and I can turn that review to an approval.

**Review by zwarm — COMMENTED** [review]
@ashiagr - All tests pass as expected. Plus I checked a few other cases not listed here (to be on the safe side). With the exception of the few minor items that @ParaskP7 called out, I am leaving my approval.

**Comment by zwarm** [comment]
> I think this is a good approach and I would be personally okay with it if @zwarm also feels the same. Just let me know and I can turn that review to an approval.

I feel good about this approach, so :shipit:

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 3 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: In 351d190

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/dynamiccards/DynamicCardsSource.kt (hunk 2 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/dynamiccards/DynamicCardsSource.kt]: In 8af4891

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartCardSource.kt (hunk 1 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartCardSource.kt:24]: 1. Actually, no changes are made to the `build(...)` invocation, it is still invoked [every time a site is selected](https://github.com/wordpress-mobile/WordPress-Android/blob/8af489104c5849661423f0070b54e9a6861df192/WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt#L154-L158).  The build function previously called `refreshData(...)` internally, we removed this call and updated the `refresh` `MediatorLiveData` value using `refresh() `call which triggers `onChanged` callback `result.addSource(refresh) { result.refreshData(...) }.` thus calling the `refreshData(...)` function as before.

    The only change here is that every time `build(...)` is invoked, `refresh` value will have the correct initial value `true` to show a loading indicator while the observing callback action is in progress. 

    Also, note that previously, on the initial fragment resume, as the screen was just built, we restricted full refresh `on FirstResume` (which internally called `MySiteRefreshSource.refresh()` updating `refresh` `MediatorLiveData` value ).  We've expanded this condition to include site switch events as well because the screen has been just built for the selected site. 

   TL;DR:  
   - `build(...)` is still invoked every time a site is selected which internally invokes`refeshData(...)`.
   - full refresh is restricted on fragment resume whenever `build(...)` is already executed and not just on an initial resume. 

   Let me know if it makes sense or I misunderstood your question. 😄 

2. `CurrentAvatarSource` is a special "site-independent" source, while the issue occurred only when a site is switched so I didn't update it. For consistency, I've updated it in 3d5ced2.

Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartCardSource.kt:24]: 👋 @ashiagr !

1. Thanks so much for this elaborate explanation! 🙏 
2. Thank you for updating, even if that was for consistency purposes only! 🙏

**Review by ParaskP7 — APPROVED** [review]
👋 @ashiagr !

Thank you for all the updates, I did took another look at them and everything LGTM, it is a 🟢 light from my side.
